modules/cli/src/main/scala/scala/cli/commands/Run.scala

```diff
@@ -75,10 +99,18 @@ object Run extends ScalaCommand[RunOptions] {
         partial = None,
         postAction = () => WatchUtil.printWatchMessage()
       ) { res =>
+        for ((process, onExitProcess) <- processOpt) {
+          onExitProcess.cancel(true)
+          ProcUtil.interruptProcess(process, logger)
+        }
         res.orReport(logger).map(_.main).foreach {
           case s: Build.Successful =>
-            maybeRun(s, allowTerminate = false)
+            val maybeProcess = maybeRun(s, allowTerminate = false)
               .orReport(logger)
+            if (options.watch.revolver)
+              processOpt = maybeProcess
+            else
+              maybeProcess.map(_._1.waitFor())
```

Comment:
NIT Maybe we should also "join" the future here, so that it doesn't run in parallel to the next watch loop iteration. Something along those lines should do
```suggestion
              for ((proc, onExit) <- maybeProcess) {
                proc.waitFor()
                try onExit.join()
                catch {
                  case _: CancellationException | CompletionException => // ignored
                }
              }
```